File: scripts/check_skill_compliance.py
#!/usr/bin/env python3
"""
Skill Compliance Checker
Validates that code follows the project's skill rules.

Usage:
    python check_skill_compliance.py <source-directory>
    python check_skill_compliance.py --staged  # Check git staged files
    python check_skill_compliance.py --file path/to/file.ts

Example:
    python check_skill_compliance.py src/
    python check_skill_compliance.py --staged --strict
"""

import argparse
import json
import re
import subprocess
import sys
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional


@dataclass
class Violation:
    file: str
    line: int
    skill: str
    rule: str
    severity: str  # CRITICAL, HIGH, MEDIUM, LOW
    message: str
    fix_suggestion: Optional[str] = None


@dataclass
class ComplianceReport:
    files_checked: int
    violations: list[Violation] = field(default_factory=list)
    errors: list[str] = field(default_factory=list)

    @property
    def critical_count(self) -> int:
        return sum(1 for v in self.violations if v.severity == "CRITICAL")

    @property
    def high_count(self) -> int:
        return sum(1 for v in self.violations if v.severity == "HIGH")

    @property
    def passed(self) -> bool:
        return self.critical_count == 0 and self.high_count == 0


def get_staged_files() -> list[str]:
    """Get list of staged files from git."""
    try:
        result = subprocess.run(
            ["git", "diff", "--cached", "--name-only", "--diff-filter=ACM"],
            capture_output=True,
            text=True,
            check=True,
        )
        return [f for f in result.stdout.strip().split("\n") if f]
    except subprocess.SubprocessError:
        return []


def get_skills() -> dict[str, dict]:
    """Load all active skills."""
    skills = {}
    skills_dir = Path(".claude/skills")

    if not skills_dir.exists():
        return skills

    for skill_dir in skills_dir.iterdir():
        if skill_dir.is_dir() and not skill_dir.name.startswith(("_", ".")):
            skill_md = skill_dir / "SKILL.md"
            if skill_md.exists():
                skills[skill_dir.name] = {
                    "name": skill_dir.name,
                    "path": skill_dir,
                    "content": skill_md.read_text(),
                }

    return skills


def get_relevant_skills(file_path: str, skills: dict) -> list[str]:
    """Determine which skills apply to a file."""
    relevant = []

    # Language-based skills
    if file_path.endswith(".ts") or file_path.endswith(".tsx"):
        if "typescript-standards" in skills:
            relevant.append("typescript-standards")
    elif file_path.endswith(".py"):
        if "python-standards" in skills:
            relevant.append("python-standards")
    elif file_path.endswith(".go"):
        if "go-standards" in skills:
            relevant.append("go-standards")

    # Location-based skills
    if "api/" in file_path or "routes/" in file_path:
        relevant.extend(["api-design", "security-hardening"])
    if "db/" in file_path or "models/" in file_path:
        relevant.extend(["database-design", "security-hardening"])
    if "components/" in file_path or "ui/" in file_path:
        relevant.extend(["ui-engineering", "accessibility-standards"])
    if "test/" in file_path or ".test." in file_path:
        relevant.append("testing-strategy")
    if "auth/" in file_path:
        relevant.extend(["auth-patterns", "security-hardening"])

    # Always check these
    relevant.extend(["error-handling", "security-hardening"])

    return list(set(relevant))


def check_typescript_standards(content: str, file_path: str) -> list[Violation]:
    """Check TypeScript-specific violations."""
    violations = []

    # Check for 'any' type
    any_pattern = r":\s*any\b"
    for match in re.finditer(any_pattern, content):
        line_num = content[: match.start()].count("\n") + 1
        violations.append(
            Violation(
                file=file_path,
                line=line_num,
                skill="typescript-standards",
                rule="No 'any' types",
                severity="HIGH",
                message="Found 'any' type - use specific type or 'unknown'",
                fix_suggestion="Replace 'any' with specific type or 'unknown'",
            )
        )

    # Check for missing return types on exported functions
    func_pattern = r"export\s+(?:async\s+)?function\s+\w+\s*\([^)]*\)(?!\s*:)(?!\s*<)"
    for match in re.finditer(func_pattern, content):
        line_num = content[: match.start()].count("\n") + 1
        violations.append(
            Violation(
                file=file_path,
                line=line_num,
                skill="typescript-standards",
                rule="Explicit return types",
                severity="MEDIUM",
                message="Exported function missing explicit return type",
                fix_suggestion="Add return type annotation",
            )
        )

    # Check for console.log
    console_pattern = r"console\.(log|warn|error|info)"
    for match in re.finditer(console_pattern, content):
        line_num = content[: match.start()].count("\n") + 1
        violations.append(
            Violation(
                file=file_path,
                line=line_num,
                skill="typescript-standards",
                rule="No console in production",
                severity="LOW",
                message="console statement found - use proper logging",
                fix_suggestion="Replace with structured logger",
            )
        )

    return violations


def check_security_hardening(content: str, file_path: str) -> list[Violation]:
    """Check security violations."""
    violations = []

    # Check for SQL injection (string concatenation in queries)
    sql_patterns = [
        r'query\s*\(\s*[`"\'].*?\$\{.*?\}',
        r'execute\s*\(\s*[`"\'].*?\+',
    ]
    for pattern in sql_patterns:
        for match in re.finditer(pattern, content, re.IGNORECASE):
            line_num = content[: match.start()].count("\n") + 1
            violations.append(
                Violation(
                    file=file_path,
                    line=line_num,
                    skill="security-hardening",
                    rule="SQL injection prevention",
                    severity="CRITICAL",
                    message="Possible SQL injection - use parameterized queries",
                    fix_suggestion="Use parameterized queries or ORM",
                )
            )

    # Check for localStorage (JWT storage)
    localstorage_pattern = r"localStorage\.(set|get)Item\s*\(\s*['\"](token|jwt|auth)"
    for match in re.finditer(localstorage_pattern, content, re.IGNORECASE):
        line_num = content[: match.start()].count("\n") + 1
        violations.append(
            Violation(
                file=file_path,
                line=line_num,
                skill="security-hardening",
                rule="JWT storage",
                severity="CRITICAL",
                message="JWT stored in localStorage - use httpOnly cookies",
                fix_suggestion="Store JWT in httpOnly secure cookie",
            )
        )

    # Check for eval
    eval_pattern = r"\beval\s*\("
    for match in re.finditer(eval_pattern, content):
        line_num = content[: match.start()].count("\n") + 1
        violations.append(
            Violation(
                file=file_path,
                line=line_num,
                skill="security-hardening",
                rule="No eval",
                severity="CRITICAL",
                message="eval() detected - major security risk",
                fix_suggestion="Remove eval, use safer alternatives",
            )
        )

    return violations


def check_error_handling(content: str, file_path: str) -> list[Violation]:
    """Check error handling violations."""
    violations = []

    # Check for generic Error
    error_pattern = r"throw\s+new\s+Error\s*\("
    for match in re.finditer(error_pattern, content):
        line_num = content[: match.start()].count("\n") + 1
        violations.append(
            Violation(
                file=file_path,
                line=line_num,
                skill="error-handling",
                rule="Custom error types",
                severity="HIGH",
                message="Generic Error thrown - use custom error type",
                fix_suggestion="Use specific error class (e.g., DatabaseError, ValidationError)",
            )
        )

    # Check for empty catch blocks
    empty_catch_pattern = r"catch\s*\([^)]*\)\s*\{\s*\}"
    for match in re.finditer(empty_catch_pattern, content):
        line_num = content[: match.start()].count("\n") + 1
        violations.append(
            Violation(
                file=file_path,
                line=line_num,
                skill="error-handling",
                rule="No empty catch blocks",
                severity="HIGH",
                message="Empty catch block - errors are silently swallowed",
                fix_suggestion="Handle error or re-throw",
            )
        )

    return violations


def check_testing_strategy(content: str, file_path: str) -> list[Violation]:
    """Check testing strategy violations."""
    violations = []

    # Check for .only in tests
    only_pattern = r"\.(only|skip)\s*\("
    for match in re.finditer(only_pattern, content):
        line_num = content[: match.start()].count("\n") + 1
        violations.append(
            Violation(
                file=file_path,
                line=line_num,
                skill="testing-strategy",
                rule="No .only/.skip in commits",
                severity="HIGH",
                message="Test has .only or .skip - will skip other tests",
                fix_suggestion="Remove .only/.skip before committing",
            )
        )

    return violations


def check_file(file_path: str, skills: dict) -> list[Violation]:
    """Check a single file for skill violations."""
    violations = []

    try:
        content = Path(file_path).read_text()
    except (FileNotFoundError, UnicodeDecodeError):
        return violations

    relevant_skills = get_relevant_skills(file_path, skills)

    # Run checks based on file type and relevant skills
    if file_path.endswith(".ts") or file_path.endswith(".tsx"):
        if "typescript-standards" in relevant_skills:
            violations.extend(check_typescript_standards(content, file_path))

    if "security-hardening" in relevant_skills:
        violations.extend(check_security_hardening(content, file_path))

    if "error-handling" in relevant_skills:
        violations.extend(check_error_handling(content, file_path))

    if ".test." in file_path or ".spec." in file_path:
        if "testing-strategy" in relevant_skills:
            violations.extend(check_testing_strategy(content, file_path))

    return violations


def print_report(report: ComplianceReport):
    """Print formatted compliance report."""
    print("\n" + "=" * 70)
    print("  SKILL COMPLIANCE REPORT")
    print("=" * 70)
    print(f"  Files checked: {report.files_checked}")
    print(f"  Violations: {len(report.violations)}")
    print(f"    🔴 Critical: {report.critical_count}")
    print(f"    🟠 High: {report.high_count}")
    print(
        f"    🟡 Medium: {sum(1 for v in report.violations if v.severity == 'MEDIUM')}"
    )
    print(f"    🟢 Low: {sum(1 for v in report.violations if v.severity == 'LOW')}")
    print("=" * 70)

    if report.violations:
        print("\n  VIOLATIONS:")
        print("  " + "-" * 68)

        # Group by severity
        for severity in ["CRITICAL", "HIGH", "MEDIUM", "LOW"]:
            severity_violations = [
                v for v in report.violations if v.severity == severity
            ]
            if severity_violations:
                icon = {"CRITICAL": "🔴", "HIGH": "🟠", "MEDIUM": "🟡", "LOW": "🟢"}[
                    severity
                ]
                print(f"\n  {icon} {severity}:")
                for v in severity_violations:
                    print(f"\n     {v.file}:{v.line}")
                    print(f"     Skill: {v.skill} | Rule: {v.rule}")
                    print(f"     Issue: {v.message}")
                    if v.fix_suggestion:
                        print(f"     Fix: {v.fix_suggestion}")

    print("\n" + "=" * 70)
    if report.passed:
        print("  ✅ ALL CHECKS PASSED")
    else:
        print("  ❌ VIOLATIONS FOUND - Fix before committing")
    print("=" * 70 + "\n")


def main():
    parser = argparse.ArgumentParser(
        description="Check code compliance with project skills"
    )
    parser.add_argument(
        "path",
        nargs="?",
        help="Directory or file to check",
    )
    parser.add_argument(
        "--staged",
        action="store_true",
        help="Check only git staged files",
    )
    parser.add_argument(
        "--file",
        help="Check specific file",
    )
    parser.add_argument(
        "--strict",
        action="store_true",
        help="Fail on any violation (including LOW)",
    )
    parser.add_argument(
        "--json",
        action="store_true",
        help="Output as JSON",
    )

    args = parser.parse_args()

    # Load skills
    skills = get_skills()
    if not skills:
        print("❌ No skills found in .claude/skills/")
        sys.exit(1)

    # Get files to check
    files_to_check = []

    if args.staged:
        files_to_check = get_staged_files()
    elif args.file:
        files_to_check = [args.file]
    elif args.path:
        path = Path(args.path)
        if path.is_file():
            files_to_check = [str(path)]
        elif path.is_dir():
            files_to_check = [str(f) for f in path.rglob("*") if f.is_file()]
    else:
        # Check current directory
        files_to_check = [str(f) for f in Path(".").rglob("*") if f.is_file()]

    # Filter to code files only
    code_extensions = {".ts", ".tsx", ".js", ".jsx", ".py", ".go", ".rs", ".java"}
    files_to_check = [
        f for f in files_to_check if any(f.endswith(ext) for ext in code_extensions)
    ]

    # Check each file
    all_violations = []
    for file_path in files_to_check:
        if Path(file_path).exists():
            violations = check_file(file_path, skills)
            all_violations.extend(violations)

    # Generate report
    report = ComplianceReport(
        files_checked=len(files_to_check),
        violations=all_violations,
    )

    if args.json:
        print(
            json.dumps(
                {
                    "files_checked": report.files_checked,
                    "violations_count": len(report.violations),
                    "passed": report.passed,
                    "violations": [
                        {
                            "file": v.file,
                            "line": v.line,
                            "skill": v.skill,
                            "rule": v.rule,
                            "severity": v.severity,
                            "message": v.message,
                        }
                        for v in report.violations
                    ],
                },
                indent=2,
            )
        )
    else:
        print_report(report)

    # Exit with error if violations found
    if args.strict and report.violations:
        sys.exit(1)
    elif not report.passed:
        sys.exit(1)
    else:
        sys.exit(0)


if __name__ == "__main__":
    main()

File: scripts/check_skill_compliance.js
#!/usr/bin/env node
/**
 * Skill Compliance Checker (JavaScript Edition)
 * Validates that code follows the project's skill rules.
 * 
 * Usage:
 *   node check_skill_compliance.js <source-directory>
 *   node check_skill_compliance.js --staged  # Check git staged files
 *   node check_skill_compliance.js --file path/to/file.ts
 * 
 * Example:
 *   node check_skill_compliance.js src/
 *   node check_skill_compliance.js --staged --strict
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

class Violation {
  constructor(file, line, skill, rule, severity, message, fixSuggestion = null) {
    this.file = file;
    this.line = line;
    this.skill = skill;
    this.rule = rule;
    this.severity = severity;
    this.message = message;
    this.fixSuggestion = fixSuggestion;
  }
}

class ComplianceReport {
  constructor(filesChecked) {
    this.filesChecked = filesChecked;
    this.violations = [];
    this.errors = [];
  }

  get criticalCount() {
    return this.violations.filter(v => v.severity === 'CRITICAL').length;
  }

  get highCount() {
    return this.violations.filter(v => v.severity === 'HIGH').length;
  }

  get passed() {
    return this.criticalCount === 0 && this.highCount === 0;
  }
}

function getStagedFiles() {
  try {
    const result = execSync('git diff --cached --name-only --diff-filter=ACM', { encoding: 'utf-8' });
    return result.trim().split('\n').filter(f => f);
  } catch (e) {
    return [];
  }
}

function getSkills() {
  const skills = {};
  const skillsDir = '.claude/skills';

  if (!fs.existsSync(skillsDir)) {
    return skills;
  }

  const items = fs.readdirSync(skillsDir);
  for (const item of items) {
    const skillPath = path.join(skillsDir, item);
    const stat = fs.statSync(skillPath);
    
    if (stat.isDirectory() && !item.startsWith('_') && !item.startsWith('.')) {
      const skillMd = path.join(skillPath, 'SKILL.md');
      if (fs.existsSync(skillMd)) {
        skills[item] = {
          name: item,
          path: skillPath,
          content: fs.readFileSync(skillMd, 'utf-8')
        };
      }
    }
  }

  return skills;
}

function getRelevantSkills(filePath, skills) {
  const relevant = [];

  // Language-based skills
  if (filePath.endsWith('.ts') || filePath.endsWith('.tsx')) {
    if (skills['typescript-standards']) relevant.push('typescript-standards');
  } else if (filePath.endsWith('.py')) {
    if (skills['python-standards']) relevant.push('python-standards');
  } else if (filePath.endsWith('.go')) {
    if (skills['go-standards']) relevant.push('go-standards');
  }

  if (skills['error-handling']) relevant.push('error-handling');

  // Location-based skills
  if (filePath.includes('api/') || filePath.includes('routes/')) {
    relevant.push('api-design', 'security-hardening');
  }
  if (filePath.includes('db/') || filePath.includes('models/')) {
    relevant.push('database-design', 'security-hardening');
  }
  if (filePath.includes('components/') || filePath.includes('ui/')) {
    relevant.push('ui-engineering', 'accessibility-standards');
  }
  if (filePath.includes('test/') || filePath.includes('.test.')) {
    relevant.push('testing-strategy');
  }
  if (filePath.includes('auth/')) {
    relevant.push('auth-patterns', 'security-hardening');
  }

  return [...new Set(relevant)];
}

function checkTypeScriptStandards(content, filePath) {
  const violations = [];
  const lines = content.split('\n');

  // Check for 'any' type
  const anyPattern = /:\s*any\b/;
  lines.forEach((line, index) => {
    if (anyPattern.test(line)) {
      violations.push(new Violation(
        filePath,
        index + 1,
        'typescript-standards',
        "No 'any' types",
        'HIGH',
        "Found 'any' type - use specific type or 'unknown'",
        "Replace 'any' with specific type or 'unknown'"
      ));
    }
  });

  // Check for missing return types on exported functions
  const funcPattern = /export\s+(?:async\s+)?function\s+\w+\s*\([^)]*\)(?!\s*:)(?!\s*<)/;
  lines.forEach((line, index) => {
    if (funcPattern.test(line)) {
      violations.push(new Violation(
        filePath,
        index + 1,
        'typescript-standards',
        'Explicit return types',
        'MEDIUM',
        'Exported function missing explicit return type',
        'Add return type annotation'
      ));
    }
  });

  // Check for console.log
  const consolePattern = /console\.(log|warn|error|info)/;
  lines.forEach((line, index) => {
    if (consolePattern.test(line) && !line.includes('//')) {
      violations.push(new Violation(
        filePath,
        index + 1,
        'typescript-standards',
        'No console in production',
        'LOW',
        'console statement found - use proper logging',
        'Replace with structured logger'
      ));
    }
  });

  return violations;
}

function checkSecurityHardening(content, filePath) {
  const violations = [];
  const lines = content.split('\n');

  // Check for SQL injection (string concatenation in queries)
  const sqlPatterns = [
    /query\s*\(\s*[`"'].*?\$\{.*?\}/,
    /execute\s*\(\s*[`"'].*?\+/,
  ];

  lines.forEach((line, index) => {
    for (const pattern of sqlPatterns) {
      if (pattern.test(line)) {
        violations.push(new Violation(
          filePath,
          index + 1,
          'security-hardening',
          'SQL injection prevention',
          'CRITICAL',
          'Possible SQL injection - use parameterized queries',
          'Use parameterized queries or ORM'
        ));
      }
    }
  });

  // Check for localStorage (JWT storage)
  const localstoragePattern = /localStorage\.(set|get)Item\s*\(\s*['"](token|jwt|auth)/i;
  lines.forEach((line, index) => {
    if (localstoragePattern.test(line)) {
      violations.push(new Violation(
        filePath,
        index + 1,
        'security-hardening',
        'JWT storage',
        'CRITICAL',
        'JWT stored in localStorage - use httpOnly cookies',
        'Store JWT in httpOnly secure cookie'
      ));
    }
  });

  // Check for eval
  const evalPattern = /\beval\s*\(/;
  lines.forEach((line, index) => {
    if (evalPattern.test(line)) {
      violations.push(new Violation(
        filePath,
        index + 1,
        'security-hardening',
        'No eval',
        'CRITICAL',
        'eval() detected - major security risk',
        'Remove eval, use safer alternatives'
      ));
    }
  });

  return violations;
}

function checkErrorHandling(content, filePath) {
  const violations = [];
  const lines = content.split('\n');

  // Check for generic Error
  const errorPattern = /throw\s+new\s+Error\s*\(/;
  lines.forEach((line, index) => {
    if (errorPattern.test(line)) {
      violations.push(new Violation(
        filePath,
        index + 1,
        'error-handling',
        'Custom error types',
        'HIGH',
        'Generic Error thrown - use custom error type',
        'Use specific error class (e.g., DatabaseError, ValidationError)'
      ));
    }
  });

  // Check for empty catch blocks
  const emptyCatchPattern = /catch\s*\([^)]*\)\s*\{\s*\}/;
  const contentWithoutNewlines = content.replace(/\n/g, ' ');
  if (emptyCatchPattern.test(contentWithoutNewlines)) {
    // Find approximate line number
    lines.forEach((line, index) => {
      if (line.includes('catch') && lines[index + 1] && lines[index + 1].trim() === '}') {
        violations.push(new Violation(
          filePath,
          index + 1,
          'error-handling',
          'No empty catch blocks',
          'HIGH',
          'Empty catch block - errors are silently swallowed',
          'Handle error or re-throw'
        ));
      }
    });
  }

  return violations;
}

function checkTestingStrategy(content, filePath) {
  const violations = [];
  const lines = content.split('\n');

  // Check for .only in tests
  const onlyPattern = /\.(only|skip)\s*\(/;
  lines.forEach((line, index) => {
    if (onlyPattern.test(line)) {
      violations.push(new Violation(
        filePath,
        index + 1,
        'testing-strategy',
        'No .only/.skip in commits',
        'HIGH',
        'Test has .only or .skip - will skip other tests',
        'Remove .only/.skip before committing'
      ));
    }
  });

  return violations;
}

function checkFile(filePath, skills) {
  const violations = [];

  try {
    const content = fs.readFileSync(filePath, 'utf-8');
  } catch (e) {
    return violations;
  }

  const content = fs.readFileSync(filePath, 'utf-8');
  const relevantSkills = getRelevantSkills(filePath, skills);

  // Run checks based on file type and relevant skills
  if (filePath.endsWith('.ts') || filePath.endsWith('.tsx')) {
    if (relevantSkills.includes('typescript-standards')) {
      violations.push(...checkTypeScriptStandards(content, filePath));
    }
  }

  if (relevantSkills.includes('security-hardening')) {
    violations.push(...checkSecurityHardening(content, filePath));
  }

  if (relevantSkills.includes('error-handling')) {
    violations.push(...checkErrorHandling(content, filePath));
  }

  if (filePath.includes('.test.') || filePath.includes('.spec.')) {
    if (relevantSkills.includes('testing-strategy')) {
      violations.push(...checkTestingStrategy(content, filePath));
    }
  }

  return violations;
}

function printReport(report) {
  console.log('\n' + '='.repeat(70));
  console.log('  SKILL COMPLIANCE REPORT');
  console.log('='.repeat(70));
  console.log(`  Files checked: ${report.filesChecked}`);
  console.log(`  Violations: ${report.violations.length}`);
  console.log(`    🔴 Critical: ${report.criticalCount}`);
  console.log(`    🟠 High: ${report.highCount}`);
  console.log(`    🟡 Medium: ${report.violations.filter(v => v.severity === 'MEDIUM').length}`);
  console.log(`    🟢 Low: ${report.violations.filter(v => v.severity === 'LOW').length}`);
  console.log('='.repeat(70));

  if (report.violations.length > 0) {
    console.log('\n  VIOLATIONS:');
    console.log('  ' + '-'.repeat(68));

    // Group by severity
    const severities = ['CRITICAL', 'HIGH', 'MEDIUM', 'LOW'];
    const icons = { CRITICAL: '🔴', HIGH: '🟠', MEDIUM: '🟡', LOW: '🟢' };

    for (const severity of severities) {
      const severityViolations = report.violations.filter(v => v.severity === severity);
      if (severityViolations.length > 0) {
        console.log(`\n  ${icons[severity]} ${severity}:`);
        for (const v of severityViolations) {
          console.log(`\n     ${v.file}:${v.line}`);
          console.log(`     Skill: ${v.skill} | Rule: ${v.rule}`);
          console.log(`     Issue: ${v.message}`);
          if (v.fixSuggestion) {
            console.log(`     Fix: ${v.fixSuggestion}`);
          }
        }
      }
    }
  }

  console.log('\n' + '='.repeat(70));
  if (report.passed) {
    console.log('  ✅ ALL CHECKS PASSED');
  } else {
    console.log('  ❌ VIOLATIONS FOUND - Fix before committing');
  }
  console.log('='.repeat(70) + '\n');
}

function main() {
  const args = process.argv.slice(2);
  
  let checkStaged = false;
  let checkFile = null;
  let sourceDir = 'src';
  let strict = false;
  let outputJson = false;

  // Parse arguments
  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--staged') {
      checkStaged = true;
    } else if (args[i] === '--file' && i + 1 < args.length) {
      checkFile = args[i + 1];
      i++;
    } else if (args[i] === '--strict') {
      strict = true;
    } else if (args[i] === '--json') {
      outputJson = true;
    } else if (!args[i].startsWith('--')) {
      sourceDir = args[i];
    }
  }

  // Load skills
  const skills = getSkills();
  if (Object.keys(skills).length === 0) {
    console.log('❌ No skills found in .claude/skills/');
    process.exit(1);
  }

  // Get files to check
  let filesToCheck = [];

  if (checkStaged) {
    filesToCheck = getStagedFiles();
  } else if (checkFile) {
    filesToCheck = [checkFile];
  } else {
    // Walk directory
    function walkDir(dir) {
      const items = fs.readdirSync(dir);
      for (const item of items) {
        const fullPath = path.join(dir, item);
        const stat = fs.statSync(fullPath);
        if (stat.isDirectory() && !item.startsWith('.') && !item.startsWith('node_modules')) {
          walkDir(fullPath);
        } else if (stat.isFile()) {
          filesToCheck.push(fullPath);
        }
      }
    }

    if (fs.existsSync(sourceDir)) {
      walkDir(sourceDir);
    }
  }

  // Filter to code files only
  const codeExtensions = ['.ts', '.tsx', '.js', '.jsx', '.py', '.go', '.rs', '.java'];
  filesToCheck = filesToCheck.filter(f => codeExtensions.some(ext => f.endsWith(ext)));

  // Check each file
  let allViolations = [];
  for (const filePath of filesToCheck) {
    if (fs.existsSync(filePath)) {
      const violations = checkFile(filePath, skills);
      allViolations.push(...violations);
    }
  }

  // Generate report
  const report = new ComplianceReport(filesToCheck.length);
  report.violations = allViolations;

  if (outputJson) {
    console.log(JSON.stringify({
      filesChecked: report.filesChecked,
      violationsCount: report.violations.length,
      passed: report.passed,
      violations: report.violations.map(v => ({
        file: v.file,
        line: v.line,
        skill: v.skill,
        rule: v.rule,
        severity: v.severity,
        message: v.message
      }))
    }, null, 2));
  } else {
    printReport(report);
  }

  // Exit with error if violations found
  if (strict && report.violations.length > 0) {
    process.exit(1);
  } else if (!report.passed) {
    process.exit(1);
  } else {
    process.exit(0);
  }
}

if (require.main === module) {
  main();
}

module.exports = { Violation, ComplianceReport, checkFile };
